### ISSUE TITLE
Document how to test a ReactN Component

### ISSUE THREAD

**Issue by michelts** [CharlesStover/reactn#48]
Hi @CharlesStover ,

I'm experimenting reactn using useGlobal hook. I'm excited with the simplicity brought by reactn against redux!

One thing I missed is a general path of how to test things up. Reactn is too fresh and there isn't too much resources out there.

I started using the useGlobal hook, and I'm testing my components using mocks. I've wrote a mock file `__mocks__/reactn.js`:

```javascript
export const useGlobal = jest.fn()
```

Suppose I have a component that apply the useGlobal like this:

```javascript
const [something, setSomething] = useGlobal('something')
```

And suppose a correponding test SomeComponent.test.js like:

```javascript
const something = 'some-value'
const setSomething = jest.fn()
useGlobal.mockImplementationOnce([something, setSomething])
const wrapper = shallow(<SomeComponent />)
wrapper.simulate('click')
expect(setSomething).toBeCalledWith('foo-bar')
expect(useGlobal).toBeCalledWith(['something']);
expect(useGlobal.mock.calls.map(call => call[0])).toEqual(['something']);
```
This can work, but I'm not sure if it is the best approach.

My question is: what do you do for test the reactn usage? Do you have any plan for test helpers? Do you see them as part of the reactn core?

**Comment by CharlesStover** [comment]
If you are wanting to spy on `useGlobal` and the resultant setter function, your implementation looks reasonably correct to me.

I do like what you've done, and I would fully support a ReactN package that collects call data. Some sort of `jest.doMock('reactn', () => require('reactn-test'))` or `require('reactn/test')` if I were to keep it in the same package.

My fear is that there are too many different ways and things to test for ReactN to build in a definitive testing package that caters to all use cases. For example, do you want to collect call data with a mock response, or do you want to collect call data and still have the global state be set? Do you want to only spy on the setter and ignore returning the state? Do you want to return the state and ignore the setter?

I am not opposed to this at all. I could see myself spearheading it, but I could also fully support any community attempt to do so as well.

I think the question should be, what do you want to have to put into your tests to make them work? Implementation details aside, what do you invision a ReactN test to look like?

**Comment by michelts** [comment]
> I think the question should be, what do you want to have to put into your tests to make them work? Implementation details aside, what do you invision a ReactN test to look like?

A helper function to configure the mock would be great. Similar to solutions to mock axios or fetch.

I can help with that! I'm not sure I can lead my own complementar package (although this could be an option). But I'm sure I can contribute to a `reactn/test`.

I'm porting two small projects to reactn. They could be my starting point to create some generalization, and maybe, the kickstart for this helper.

I will come back here again with my results, as far as I have something, ok?

**Comment by matthew-davis** [comment]
@michelts if you have a working example of a test I'd love to see it. I have a use case for reactn in my current project, but cannot currently include it without the ability to test it. Two issues I had with your above code is that I have never actually seen __mocks__ work. I'm sure that's a larger issue with me needing to spend more time with mocks. But in an move of desperation I created a __mocks__ folder at root next to the node_modules folder and included reactn.js with the code:
```
// __mocks__/reactn.js
export const useGlobal = jest.fn();
```
Then I had to make a slight variation to your code above as mockImplementationOnce requires a function not an array (I'm sure you were just throwing together an example though) I also had a component that had a child element of a button, so its neither here nor there but I had to find the button before I clicked it:
```
// src/components/__tests__/SomeComponent.test
import * as React from 'react';
import {SomeComponent} from '../SomeComponent';
import {shallow} from 'enzyme';
import {useGlobal} from 'reactn';

describe('Testing a Test', () => {
  it('Calling SetSomething', () => {
    const something = 'some-value';
    const setSomething = jest.fn();
    useGlobal.mockImplementationOnce(() => [something, setSomething]);
    const wrapper = shallow(<SomeComponent />);
    const somethingButton = wrapper.find('#somethingButton').simulate('click', {});
    expect(setSomething).toBeCalledWith('foo-bar');
    expect(useGlobal).toBeCalledWith(['something']);
    expect(useGlobal.mock.calls.map(call => call[0])).toEqual(['something']);
  });
});
```
Any pointers would be helpful to allow for testing as I'm very keen to include this in our project.

**Comment by michelts** [comment]
@matthew-davis you don't need the `__mocks__` folder at all. You can just mock using jest.mock in each test file. But your [manual mock](https://jestjs.io/docs/en/manual-mocks) is ok.

I wrote mockImplementationOnce but in fact I mean mockReturnValueOnce, although you can use your approach too.

Your test code isn't wrong. I personally won't test useGlobal was called, if you test `something` appear in your output and test `setSomething` is executed, useGlobal had to been called.

I have been using reactn since 10 days ago and I feel very comfortable working with hooks. I've tried component based operation and it is also cool, but it very difficult to test. I need to use the jest.doMock, resetModules and things that make testing more complex.

I decided to keep functional components, that are much easier to test.

I wrote an [article about how to test functional components that uses reactn' useGlobal](https://link.medium.com/TcsIjYtMgV), mainly to share with my team. Maybe it can help.

**Comment by CharlesStover** [comment]
Thank you @michelts for the article. I am working on 2.0 in the `typescript` branch. It is 2.0 because it will change reducer syntax (`this.dispatch.reducerName` instead of `this.global.reducerName`, and `useGlobalReducer('reducerName')` instead of `useGlobal('reducerName')`), but besides that, everything else should function exactly the same.

An upside to being 2.0 is I can add all sorts of features without worrying about impacting current implementations. If you have any suggestions for simplifying testing, now is a great opportunity to get that in.

I'll definitely read over your article and @matthew-davis's comment to see if I can find any testing patterns that can be simplified or features I can add to aid in the process.

I think I can even put test-specific features behind a `NODE_ENV === 'test'` flag so that they don't impact production or dev environments. :)

Thank you both for your contributions to this idea. I really support it, and community feedback is something I definitely need to make ReactN as versatile as I want it to be.

**Comment by michelts** [comment]
@CharlesStover Test useGlobal is very simple using jest mocks and even with the 2.0 changes, it is still just matter of mock.

The problem resides in testing class based components. 

I wrote a draft of a helper function:

```javascript
const ReactNImpl = (mockedGlobal, setGlobalFn) => () => (WrappedComponent) => {
  class ReactNMockedComponent extends WrappedComponent {
    static displayName = `${(WrappedComponent.displayName || WrappedComponent.name)}-ReactNMock`;

    get global() {
      return mockedGlobal;
    }

    setGlobal = setGlobalFn
  }

  return ReactNMockedComponent;
};

export default ReactNImpl;
```

I can use it like this:

```javascript
describe('Component', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it('some test', () => {
    const initialGlobal = { ...myGlobalCameThroughFactories };
    const setGlobal = jest.fn();
    jest.doMock('reactn', ReactNImpl(initialGlobal, setGlobal));
    const Component = require('../component').default;
    const wrapper = shallow(<Component />);
    ...
  });
});
```

What bugs me here is the need to reimport Component using `require` in each test. I can do it in a getWrapper function and generalize things, but anyway, it seems to me too much thing to, do compared with functional components - or maybe I didn't realize a better way to do it :sweat_smile:.

It would be simpler if I use the withGlobal HOC to translate the global state to props, then it's just matter of mock, similar to the useGlobal, although useGlobal (and hooks) sound much simpler to me.

I'm not interested in writing and mainiting an external helper (because I think it is not needed). I'm even not sure the helper I wrote cover all the paths. Anyway, if you want to include it in Reactn, I can help.

But I can contribute with documentation, explaining how to test useGlobal and giving general guides on how to test class based components. Can I start this way?

**Comment by CharlesStover** [comment]
>The problem resides in testing class based components.

What problem did you run into with testing class components? I would think they would be easier.

```JavaScript

// render with your renderer or choice
const instance = renderer.shallow(<ReactNComponent />);

// one or the other
const spy = jest.spyOn(instance, 'setGlobal');
const spy = sinon.spy(instance, 'setGlobal');

// action
instance.handleButtonClick('test');

// expectation
// -- jest syntax
expect(spy).toHaveBeenCalledWith({ newValue: 'test' });
// -- sinon syntax
expect(spy.toHaveBeenCalledOnceWithExactly({ newValue: 'test' })).to.equal(true);
```

Component instances can be mocked using the current Jest and Sinon implementations because the methods exist on an object.

Mocking `useGlobal` is definitely a consideration, since functional components do not have instances. I want the test functionality for that to match how `useState` is meant to be tested as closely as possible.

**Comment by michelts** [comment]
I will try the spyOn approach. I'm not very familiar with it.

Anyway, I need to be able to configure the global attribute (and now it's clear I can do it just updating the instance) and mock the setGlobal.

I choose the harder path, my bad :)

> I want the test functionality for that to match how useState is meant to be tested as closely as possible.

I have been trying to test builtin hooks but I don't think enzyme users can enjoy it right now. Anyway, I don't see their tests different from a class based stateful component test:

```javascript
const wrapper = shallow(...)
wrapper.simulate('click') // internally, the state is updated
expect(wrapper.find('something')).toEqual(updatedThing)
```

This way, we hide implementation details. It is different from what I did in my examples and different from what I do with redux. I used to test:

* the component updates the global state
* different global state scenarios impact component behavior

And I think this is the right approach. I'm using a global state because the component that respond to the global state not necessarily trigger global state updates.

I will try the spyOn approach with class based component and bring more samples.

Thanks!

**Comment by michelts** [comment]
I'm not sure spyOn would fit all the cases. I'm doing setGlobal in the componentDidMount, instance would be mounted before I can spy his methods or change his global attribute. I need them to be set at mount time.

The helper function I wrote do it, but besides exec the doMock in each test, I need also to require again the component. The final test code seems odd this way.

**Comment by CharlesStover** [comment]
Great use case. I had to search for how others are testing `componentDidMount`. This Stackoverflow post should work: https://stackoverflow.com/questions/43245040/using-jest-to-spy-on-method-call-in-componentdidmount

`setGlobal` for class Components is defined on the prototype, so it can be spied on.

```JavaScript
const setGlobalSpy = jest.spyOn(MyComponent.prototype, 'setGlobal');
renderer.mount(<MyComponent />);
expect(setGlobalSpy).toHaveBeenCalledWith({ initial: 'state' });
```

Let me know if this doesn't match your use case or what other use cases are still unhandled or need documentation. I'll definitely include a testing README for 2.0.

**Comment by michelts** [comment]
> This Stackoverflow post should work

Perfetct! Thank you!

This approach resolve the problem. I can mock the global state the same way:

```javascript
import React from 'react';
import { shallow } from 'enzyme';
import Counter from './Counter';

describe('Counter component', () => {
  it('should set count to zero when mounted', () => {
    const setGlobal = jest.spyOn(Cards.prototype, 'setGlobal');
    shallow(<Counter />);
    expect(setGlobal).toBeCalledWith({ count: 0 });
  });

  it('should show the global count value', () => {
    jest.spyOn(Cards.prototype, 'global', 'get').mockReturnValue({ count: 3 });
    const wrapper = shallow(<Counter />);
    expect(wrapper.text()).toEqual('Count: 3');
  });
});
```

The key here is to return a value from the spyOn call, and to spyOn the global getter (using the `get` access type).

As long as reactn expose the global state through this getter, the test code will work.

> or what other use cases are still unhandled

I think everything is covered now. Do you want help with the docs? I can make a pull request. What is your deadline?

**Comment by CharlesStover** [comment]
I don't think I'll need help with the docs. I'll throw it into the README as I finish writing tests for 2.0. You are more than welcome to write Medium articles about your experiences. I definitely appreciate any publicity. 🙂

I think spying on the class components is intuitive and aligns with current React testing methods for non-ReactN Components, so I am happy with that.

I have never written a test for a functional Component with hooks. I cannot verify that it aligns with current React testing methods for non-ReactN Components. If it does not, let me know, because I want to make it so.

I am also open to feedback on new features unique to ReactN that make testing easier. I would be happy to reduce any boilerplate caused by ReactN. But I'm at a loss for what those features could possibly be or if there is even any boilerplate separate from what is required of testing React in the first place.

**Comment by michelts** [comment]
> I definitely appreciate any publicity

Count on me for that :)

**Comment by bovas85** [comment]
Was this removed from the docs or milestone? It'd help me and my org to see the docs in the official Github for testing with jest. I can refer to this issue here but it's better to have something set in stone. At least some examples here and there

**Comment by CharlesStover** [comment]
It is still a milestone, just lower priority. Priority goes to features and bug fixes that are blocking other developers or people from reaching their goals. While I want thorough documentation for the package, this is more of a nice-to-have.

**Comment by bovas85** [comment]
Thanks @CharlesStover, if you have a draft I could help contribute to it.
I think a general idea of the direction can be inferred from this page

**Comment by CharlesStover** [comment]
This issue is the only current source of discussion for testing; I have no draft. The `docs` directory of this repository is a work-in-progress of documentation, and my efforts towards documentation have been migrating the README to the `docs` SPA. That said, I am not wholly impressed by the SPA, so contributing to a `.md` file will probably be the fastest way to launch a new section of documentation.

**Comment by bovas85** [comment]
Sounds good, I'll work on the md once I have a minute

**Comment by bovas85** [comment]
Sorry, currently have no bandwidth to work on this as we don't use it at work

**Comment by roni-castro** [comment]
Why doesn't this spy work? 

```
import * as Reactn from 'reactn'
const setGlobal = jest.spyOn(Reactn, 'setGlobal')
```

**Comment by michelts** [comment]
Hi @roni-castro 

AFAIK you can't use external methods in manual mocks. I mean, you probably can't use setGlobalMock this way. But you can import your setGlobal function and you will have the mocked one in hands. For instance:

```
import { setGlobal, getGlobal } from 'reactn';

jest.mock('reactn', () => ({
    getGlobal: jest.fn(),
    setGlobal: jest.fn()
}))

expect(setGlobal).toBeCalledWith(...)
getGlobal.mockReturnValue(...)
```

**Comment by CharlesStover** [comment]
I'd like to preemptively clarify that spying on `setGlobal` as imported through `'reactn'` will have no impact on `setGlobal` as returned by `useGlobal()` or the `setGlobal` method on class components. Those are three different functions that would each need to be mocked independently.

It's entirely up to your transpiler (e.g. TypeScript version and configuration or Babel plugins/version and config) whether or not spying on `import * as module` actually mocks the module. I can't provide strong solutions here, as it's not within the scope of this package.

I have some less-than-ideal boilerplate that has worked for me in some projects where I could not tweak the transpiler settings. This code may work for you:

```
// app.tsx
import * as ReactN from 'reactn';
const { setGlobal } = ReactN;

setGlobal(INITIAL_STATE);
```

```
// app.test.ts
import * as ReactN from 'reactn';
const MOCK_SET_GLOBAL = jest.spyOn(ReactN, 'setGlobal');
import { All, Other, Imports } from './app';
```

In this example, two things of note (that again are dependent on your transpiler, so I cannot give strong solutions):
* The non-test file also imports `* as module` before destructuring the values, as opposed to destructuring while importing.
* The test file imports `* as module` and spies on the functions before importing any other values.

While these implementation details are not needed for mocking to work in all projects (hopefully mocking should just work out of the box), I hope they help unblock you.
